navis/nbl/smat.py

```diff
@@ -201,43 +225,82 @@ def with_bin_counts(self, bin_counts: List[int]):
         self.digitizers = None
         return self
 
-    def _dotprop_keys(self) -> Sequence[DotpropKey]:
-        """Get all indices into dotprops instance member"""
+    def _object_keys(self) -> Sequence[DotpropKey]:
+        """Get all indices into objects instance member."""
         try:
-            return self.dotprops.keys()
+            return self.objects.keys()
         except AttributeError:
-            return range(len(self.dotprops))
+            return range(len(self.objects))
 
     @property
     def nonmatching(self) -> List[DotpropKey]:
-        """Indices of nonmatching set of neurons"""
+        """Indices of nonmatching set of neurons."""
         if self._nonmatching_list is None:
-            return list(self._dotprop_keys())
+            return list(self._object_keys())
         return self._nonmatching_list
 
     def _yield_matching_pairs(self) -> Iterator[Tuple[DotpropKey, DotpropKey]]:
-        """Yield all index pairs within all matching pairs"""
+        """Yield all index pairs within all matching pairs."""
         for ms in self.matching_lists:
             yield from yield_not_same(permutations(ms, 2))
 
-    def _yield_nonmatching_pairs(self) -> Iterator[Tuple[DotpropKey, DotpropKey]]:
-        """Yield all index pairs within nonmatching list"""
-        # todo: this could be much better, use meshgrid or shuffle index arrays
+    def _yield_nonmatching_pairs_greedy(self, rng=None) -> Iterator[Tuple[DotpropKey, DotpropKey]]:
+        """Yield all index pairs within nonmatching list."""
         return yield_not_same(permutations(self.nonmatching, 2))
 
+    def _yield_nonmatching_pairs_batched(self) -> Iterator[Tuple[DotpropKey, DotpropKey]]:
```

Comment:
@clbarnes: not sure where our original thread on this went but I have now implemented two methods for randomly drawing non-matches:
1. "greedy" is the original method, i.e. it will generate all possible matches and then shuffle them before drawing the necessary number
2. "batched" (default) is using a generator that creates smaller batches of randomly drawn non-matches to avoid generating more than absolutely necessary